---
title: "Model Zero-Inflated Microbiome"
author: "David Xin Zhao"
date: "Last edited `r format(Sys.time(), '%d %B %Y')`"
knit: (function(inputFile, encoding) {
      out_dir <- 'docs';
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_file=file.path(dirname(inputFile), out_dir, 'index.html'))})
output:
  html_document:
    theme: cosmo
    highlight: pygments
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    collapsed: FALSE
    number_sections: TRUE
    fig_width: 7
    fig_height: 6
    fig_caption: TRUE
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load useful libraries 
```{r r library, message=FALSE, warning=FALSE, results='hide'}

library(tidyverse) # data manipulation and visualizations 
library(pscl)  # implement ZIP, ZINB, ZHP, and ZHNB models 
library(Maaslin2) # provides microbiome and meta data sets 
library(readr)
library(vegan)
library(lmtest)
library(nonnest2)

``` 

## Basic principle of zero-inflated models 
- Types of zero in microbiome abundance data
        1. Rounded zero
        2. Sampling zero 
        3. Structural zero: main source of excessive zero which Poisson regression cannot fit well 
        
- Mixture models: In mixture models, the rounded zeros are explicitly modeled with a distribution, i.e., the binomial distribution. The structural zeros are modeled as the probability of expecting a zero under the considered distribution, i.e., the negative binomial or the Poisson distribution; and the nonzero values are analyzed as the rest of the data. \ 
Zero-inflated Poisson (ZIP) and zero-inflated negative binomial (ZINB) belongs to this category. \ 

- Two-part models: The two-part model consists of two parts: first, the data are considered as zeros versus non-zeros and the probability of a zero value is modeled by a binomial distribution; and second, the non-zero values are analyzed via a truncated Poisson or truncated negative binomial models.\ 
Zero-hurdle Poisson (ZHP) and zero-hurdle negative binomial (ZHNB) belongs to this category. 

- ZIP/ ZHP focuses on the violation of the Poisson by the population heterogeneity in the presence of structural zeros; while ZINB/ ZHNB address both sources of heterogeneity, namely over dispersion and excessive zeros. 



## Procedure at a glance 
1. Load abundance data and prepare analysis abundance 
2. Investigate outcome distribution zeros 
3. Create offset for Poisson/ NB models (to calculate ratios)
4. Create a formula for fitting ZIP/ ZINB and ZHP/ ZHNB 
5. Fit ZIP/ ZINB and ZHP/ ZHNB
6. Select the optimal model
7. Interpret models' output 


```{r set working dir, echo=FALSE}

setwd("C:/Users/17803/Documents/RStuodio-link-GitHub/zero-inflated-microbiome-model")

```


## Step 1 Preprocess data for downstream modeling 

The 16S abundance data and metadata was retrieved from the [Inflammatory Bowel Disease Multi'omics Database](https://ibdmdb.org/tunnel/public/HMP2/16S/1806/products).  

```{r microbiome count data}

# read in 16S abundance/ count data in HMP2 database 
untar("hmp2_abund.tar", exdir = getwd()) # unzip a .tar file 

abund_raw <- read_tsv("taxonomic_profiles.tsv.gz", 
                      show_col_types = FALSE) %>% 
        as.data.frame() 

``` 

Read in the metadata in `.csv` format. 
```{r metadata dataframe}
# read in the meta data 
meta_raw <- read.csv("hmp2_metadata.csv")

```

Subset identifiers, environmental/ clinical variables of my research interest, including `Project`, `External.ID`, `Participant.ID`, `week_num`, `biopsy_location`, `diagnosis` and `Age.at.diagnosis`. 

```{r subset env variables}

meta_raw_sub <- meta_raw %>% 
        select(Project, 
               External.ID,
               Participant.ID,
               diagnosis,
               week_num, 
               biopsy_location,
               Age.at.diagnosis) 

```

Merge 16S abundance data and subset metadata. 
```{r merge 16S and metadata}

# reshape abundance data 
abund_raw_id <- abund_raw %>% 
        gather(key="sample_id",
               value = "read",
               -c("#OTU ID", "taxonomy")) 

# join two data frames by "id" = "External.ID" 
abund_meta <- abund_raw_id %>% 
        inner_join(meta_raw_sub, 
                   by = c("sample_id" = "External.ID"))

```


The microbiome 16S abundance data set contains data points from `r length(unique(abund_meta$Participant.ID))` participants; the biopsy were collected from 7 different locations, including `Rectum`, `Ileum`, `Sigmoid Colon`, `Descending (left-sided) colon`, `Cecum`, `Ascending (right-sided) colon`, and `Transverse colon`. \ 
For the simplicity purpose, I focused on rectum biopsy data for the subsequent analysis. 
```{r extract only rectum biopsy data points} 
abund_meta_rectum <- abund_meta %>% 
        filter(biopsy_location == "Rectum") 

```


Collapse taxonomy at the genus level, namely similarity equals or higher than 97% 
```{r collapse the taxonomy at the genus level}

abund_meta_rectum2 <- abund_meta_rectum %>%
        mutate(genus = str_split_fixed(abund_meta_rectum$taxonomy, "; __", 6)[,6]) %>%  
        mutate(genus_rep = case_when(genus == "g" ~ "unclassified", 
                                     genus != "g" ~ genus)) %>%
        select(-genus)

```


Select columns, `sample_id`, `Project`, `Participant.ID`, `read`, `diagnosis` and `genus_rep` from the above data frame. And then calculate total reads for each individual as the offset. 

```{r offset calculation}

abund_meta_rectum3 <- abund_meta_rectum2 %>% 
        select(sample_id, Participant.ID, read, diagnosis, genus_rep) %>%
        rename(subject_id = Participant.ID)

df_total_reads <- abund_meta_rectum3 %>% 
        group_by(sample_id) %>% 
        summarise(lib_size = sum(read)) %>%
        ungroup() 

abund_meta_rectum4 <- abund_meta_rectum3 %>% 
        left_join(df_total_reads, by="sample_id") 


# convert to the factor variable with new levels 
abund_meta_rectum4$diagnosis <- factor(abund_meta_rectum4$diagnosis,
                                       levels = c("nonIBD", "CD", "UC"))


# abund_meta_rectum3 %>% 
#         group_by(sample_id) %>%
#         summarise(lib_size= sum(read)) %>% 
#         arrange(lib_size) %>%
#         ggplot(aes(x=1, y=lib_size)) +  # visualize distribution of variations in library sizes 
#         geom_jitter()+
#         scale_y_log10() 

```
The sampling depth varied by samples which likely affect hypothesis analysis. \ 
I consider to choose a best depth threshold and to conduct rarefaction, following the [instruction](https://www.youtube.com/watch?v=c7H8jLjTxSE) introduced by Dr. Pat Schloss. \ 
The optimal sampling depth threshold is determined according to the good's coverage statistics. 

```{r determin best sampling depth}

# abund_meta_rectum3 %>% 
#         group_by(sample_id) %>% 
#         summarise(singles = sum(read == 1),
#                   lib_size = sum(read)) %>% 
#         mutate(goods_cov = 100*(1-singles/lib_size)) %>%  # calculate the good's coverage 
#         ggplot(aes(x= lib_size, y=goods_cov))+ # plot sampling size by goods coverage 
#         geom_point() + 
#         geom_hline(yintercept = 97, color = "red")  # good's coverage = 97 percentage 
# 
# 
# abund_meta_rectum3_goods <- abund_meta_rectum3 %>% 
#         group_by(sample_id) %>% 
#         summarise(singles = sum(read == 1),
#                   lib_size = sum(read)) %>% 
#         mutate(goods_cov = 100*(1-singles/lib_size)) 
# 
# 
# abund_meta_rectum3_goods %>% 
#         filter(goods_cov >= 97) %>% 
#         arrange(lib_size)  # sampling depth threshold = 286 reads
# 
# 
# # identify ids for samples below the minimal sampling depth 
# sample_drop <- abund_meta_rectum3_goods %>% 
#         filter(goods_cov <97)  %>% 
#         pull(sample_id)
# 

```


## Step 2 Investigate outcome distribution zeros 

Subset samples with zero percentage higher ranging from 50% to close 100%. Apply models specializing in the zero-inflated data to the sub-samples. 
```{r histogram of zero distribution in genus}

df_zero_percent <- abund_meta_rectum4 %>% 
        filter(genus_rep != "unclassified") %>% 
        group_by(genus_rep) %>% 
        summarise(zero_percent = mean(read == 0)) 

df_zero_percent$zero_cut <- cut(df_zero_percent$zero_percent, breaks = 4)  # cut zero_percentage into four groups 

levels(df_zero_percent$zero_cut)

# subset variables for the genus with many zero 
abund_meta_rectum5 <- abund_meta_rectum4 %>% 
        left_join(df_zero_percent, by= "genus_rep") %>% 
        arrange(desc(zero_percent)) 

```


Create histogram for representative genus, `_Clostridium_innocuum_group` with excessive zeros 
```{r histogram for zero distribution}

abund_meta_rectum5 %>% 
        filter(zero_percent != 1) %>% 
        filter(zero_cut == "(0.506,0.753]") %>% 
        head(10)


```

Use `dpois` to estimate probability of zeros, `zero_prop_poisson` assuming that sample follows Poisson distribution. The estimated frequency of zero `zero_freq_poisson` is calculated based on `zero_prop_poisson`. 

In addition, calculate empirical zero frequency and average in the sample. 

We can tell whether excessive zero exist when empirical zero frequency is (or not) much higher than estimated one under the Poisson distribution; we can also tell whether over-dispersion exist when empirical average not close to variance. 


```{r mean and zero count}

zero_summary_innocuum <- abund_meta_rectum5 %>%
  filter(genus_rep == "_Clostridium_innocuum_group") %>% 
  group_by(diagnosis) %>% 
  summarise(sample_size = n(),
            zero_read = sum(read == 0),
            average_read = round(mean(read), 1),
            variance = round(var(read), 1)) %>% 
  mutate(zero_prop_poisson = round(dpois(0, average_read),2)) %>% 
  mutate(zero_freq_poisson = zero_prop_poisson*sample_size)

# print the table 
zero_summary_innocuum  


```
Excessive zero exist in abundance distribution of the taxa `_Clostridium_innocuum_group`, since empirical zero frequency is much higher than estimated zero frequency under Poisson distribution. And, a large proportion of empirical zero is the structural zero. 

In addition, huge difference between empirical average and variance exist in the sample, indicating over-dispersion. 

Therefore, zero-inflated or zero-hurdle negative binomial might be a suitable choice. 


```{r histogram plot}
# use "_Clostridium_innocuum_group" as an example for plotting histogram 

abund_meta_rectum5$diagnosis <- factor(abund_meta_rectum5$diagnosis,
                                          levels = c("nonIBD", "CD", "UC"))

hist_innocuum <- abund_meta_rectum5 %>% 
        filter(genus_rep == "_Clostridium_innocuum_group") %>% 
        ggplot(aes(x=read))+
        geom_histogram(aes(fill = diagnosis), alpha=0.5, binwidth = 10)+
        xlab("Observed read values")+
        ylab("Frequency")+
        coord_cartesian(ylim = c(0, 10), xlim=c(0,100))+
        facet_wrap(vars(diagnosis))


df_innocuum_label <- zero_summary_innocuum %>% 
  mutate(sample_size_lab = paste("Sample size", sample_size, sep = ": ")) %>% 
  mutate(zero_read_lab = paste("Zero count", zero_read, sep = ": ")) %>% 
  mutate(average_read_lab = paste("Average", average_read, sep = ": ")) 

hist_innocuum +
  geom_text(data = df_innocuum_label,
            aes(label = sample_size_lab), x=40, y=9.5, hjust=0, size=3.5) +
  geom_text(data = df_innocuum_label,
            aes(label = zero_read_lab), x=40, y=9, hjust=0, size=3.5)+
  geom_text(data = df_innocuum_label,
            aes(label = average_read_lab), x=40, y=8.5, hjust=0, size=3.5)+
  geom_vline(data= df_innocuum_label, aes(xintercept = average_read), color="blue", linetype="dashed")


```

The histogram illustrates that empirical zero frequency and means vary by the diagnosis status. 

## Step 3 Create offset for poisson/ NB models to calculate ratios
The total count reads is used to calculate offset. The offset will be adjusted as a covariate in the model later to ensure microbiome response is relative abundance instead of count data. 

This step is critical in fitting linear mixed effects modesl in microbiome study. 

```{r create offet for poisson and NB models}
abund_meta_rectum6 <- abund_meta_rectum5 %>%
        mutate(Offset =  log(lib_size)) %>% 
        filter(genus_rep == "_Clostridium_innocuum_group") 
        
head(abund_meta_rectum6$Offset)  # view first rows 
```


## Step 4 Create a formula for fitting ZIP/ ZINB and ZHP/ ZHNB
The formula can be used to specify both components of the model pertaining to count and structural zero models. 

To illustrate more features of zero-inflated and zero-hurdle models and to compare these features, in the following analysis, I choose the same covariate for both count and binomial distribution, but only set offset argument to count component. 

The response variable is `read` (abundance) of the taxa `_Clostridium_innocuum_group`, and the independent (explanatory) variable is `diagnosis` which is a multi-level, categorical variable. The [offset](https://www.youtube.com/watch?v=EyKba_sOp0k&list=PLLTSM0eKjC2cYVUoex9WZpTEYyvw5buRc&index=9) term is logarithm value of total read in individual samples. 

```{r establish the formula}

f_model <- formula(read ~ diagnosis + offset(Offset) | diagnosis)  

```
There two components in the formula. The first component is Poisson or Negative Binomial regression; and the second one the binomial regression. In particular, `read ~ diagnosis + offset(Offset)` is for the first component, while `read ~ diagnosis` is for binomial regression. 

## Step 5 Fit ZIP/ ZINB and ZHP/ ZHNB 

Fit a ZIP model for the representative taxa `_Clostridium_innocuum_group` as the function of `diagnosis`, indicating `nonIBD` versus `UC` versus `CD`. 

Use `zeroinfl()` function of `pscl` package to fit ZIP model. 
```{r fit ZIP mdoel}
ZIP_innocuum <- zeroinfl(formula = f_model, 
                         dist = "poisson",
                         link = "logit",
                         data = abund_meta_rectum6) 

```

In above formula, the `dist` option specifies the distribution for the count data. Currently, Poisson negative binomial, and geometric distribution are available in `pscl` package. 

The `link = logit` option specifies the logistic link for the structural zeros versus non-structural zeros (the sampling zeros plus the positive counts). 

The `offset` term (the log of the total reads in a given sample) is used here to allow for a comparison in the relative abundance (and not absolute counts) among groups. 


```{r model summary}
summary(ZIP_innocuum)

```

The row `BFGS optimization` refers to the number of optimal iterations. `diagnosisCD` and `diagnosisUC` are significant for poisson with log link function alone. Note ZIP model assumes that no over-dispersion in the non-zero count data. 

`Log-likelihood` measures how well the model fits data points. 


Below fit ZINB model alternatively. 
```{r ZINB model}
ZINB_innocuum <- zeroinfl(formula = f_model,
                          dist = "negbin",
                          link="logit",
                          data= abund_meta_rectum6) 

# print ZINB output 
summary(ZINB_innocuum)
```
The `Log(theta)` value is significant, indicating that abundance are over-dispersed. 


Use the same formula and offset as fitting ZIP and ZINB models above to fit ZHP and ZHNB models using the function `hurdle()` from the `pscl` package. 
```{r ZHP model}
ZHP_innocuum <- hurdle(formula = f_model,
                       dist = "poisson",
                       data = abund_meta_rectum6) 

summary(ZHP_innocuum)
```


Alternatively, fit ZHNB model with identical formula and offset. 
```{r ZHNB model}

ZHNB_innocuum <- hurdle(formula = f_model,
                        dist = "negbin",
                        data = abund_meta_rectum6) 

summary(ZHNB_innocuum)
```



## Step 6 Select the optimal model 

Compare four models, `ZIP_innocuum`, `ZINB_innocuum`, `ZHP_innocuum`, and `ZHNB_innocuum` based on alternative statsitics (i.e., [likelihood ratio](https://www.youtube.com/watch?v=tDvwpET-nJQ&list=PLLTSM0eKjC2cYVUoex9WZpTEYyvw5buRc&index=12), and [AIC](https://www.youtube.com/watch?v=tDvwpET-nJQ&list=PLLTSM0eKjC2cYVUoex9WZpTEYyvw5buRc&index=12)). 

Note that likelihood ratio test is suitable for nested models alone; whereas AIC is suitable for both nested and non-nested models. 

Use `lmtest` package for statistics computation. 

```{r LRT}
# likelihood 
lmtest::lrtest(ZIP_innocuum, ZINB_innocuum) 

lmtest::lrtest(ZHP_innocuum, ZHNB_innocuum)


```
According to likelihood ratio tests,  `ZINB_innocuum` outperforms `ZIP_innocuum` since the former has significantly smaller residuals (deviance) than the later one. 

`ZHNB_innocuum` outperforms `ZHP_innocuum`, since the former has significantly smaller residuals (deviance) than the later one. 

```{r AIC}
# AIC
t(AIC(ZIP_innocuum, ZINB_innocuum, ZHP_innocuum, ZHNB_innocuum)) # AIC smaller is better 

```
According to AIC,  `ZINB_innocuum` seems the best model in the present case, since it has the smallest AIC relative to other three counterparts. `ZHNB_innocuum` was the second best model. 


## Step 7 Interpret the model coefficients 
Use ZINB_innocuum model as the example to present how to interpret coefficients in a reasonable way. 

```{r print summary output}
summary(ZINB_innocuum)

```

1. Count model coefficients: 
- The average relative abundance (excluding structural zero) of the taxa `_Clostridium_innocuum_group` at the baseline (`nonIBD`) is `r exp(-6.2744)`.  
- The average relative abundance (excluding structural zero) in `diagnosisUC` group is `r 100*(1-exp(-1.9110))`% lower relative to the baseline level. 
- Log(theta) is significant, indicating that over-dispersion exists. 

2. Zero-inflation model coefficients: 
- p values for the intercept, and coefficients are not significant, indicating there is no significant difference in structural zeros among three diagnosis status. 



## Final remarks 
Although ZINB model outperforms ZHNB model in fitting abundance of `_Clostridium_innocuum_group`, we should note that the zero-inflated model, however assume the existence of a latent structural-zero class. If such a class does not seem justified, a hurdle model should be preferred to use fo the easier interpretation of zeros.[@p.480, Statistical Analysis of Microbiome Data with R]

The authors recommend in the book, Statistical Analysis of Microbiome Data with R that model selection should be based on:
1. Model fitting (to choose better-fitted models)
2. Conceptual appropriateness (the chosen models should be conceptual interpretative)
3. Parsimony (i.e., [Occam's razor](https://en.wikipedia.org/wiki/Occam%27s_razor), given all the criteria met, the simplest is selected as the best model) 




