---
title: "model-zero-inflated-microbiome"
author: "David Xin Zhao"
date: "2022-11-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Load useful libraries 
```{r r library, message=FALSE, warning=FALSE, results='hide'}

library(tidyverse) # data manipulation and visualizations 
library(pscl)  # implement ZIP, ZINB, ZHP, and ZHNB models 
library(Maaslin2) # provides microbiome and meta data sets 
library(readr)
library(vegan)

``` 

## Basic principle of zero-inflated models 
- Types of zero in microbiome abundance data
        1. Rounded zero
        2. Sampling zero 
        3. Structural zero 
        
- Mixture models: In mixture models, the rounded zeros are explicitly modeled with a distribution, i.e., the binomial distribution. The structural zeros are modeled as the probability of expecting a zero under the considered distribution, i.e., the negative binomial or the Poisson distribution; and the nonzero values are analyzed as the rest of the data. \ 
Zero-inflated Poisson (ZIP) and zero-inflated negative binomial (ZINB) belongs to this category. \ 

- Two-part models: The two-part model consists of two parts: first, the data are considered as zeros versus non-zeros and the probability of a zero value is modeled by a binomial distribution; and second, the non-zero values are analyzed via a truncated Poisson or truncated negative binomial models.\ 
Zero-hurdle Poisson (ZHP) and zero-hurdle negative binomial (ZHNB) belongs to this category. 

- ZIP/ ZHP focuses on the violation of the Poisson by the population heterogeneity in the presence of structural zeros; while ZINB/ ZHNB address both sources of heterogeneity, namely over dispersion and excessive zeros. 



## Procedure at a glance 
1. Load abundance data and prepare analysis abundance 
2. Investigate outcome distribution zeros 
3. Create offset for Poisson/ NB models (to calculate ratios)
4. Create a formula for fitting ZIP/ ZINB and ZHP/ ZHNB 
5. Fit ZIP/ ZINB and ZHP/ ZHNB
6. Select the optimal model
7. Interpret models' output 


```{r set working dir, echo=FALSE}

setwd("C:/Users/17803/Documents/RStuodio-link-GitHub/zero-inflated-microbiome-model")

```


## Step 1 Preprocess data for downstream modeling 

The 16S abundance data and metadata was retrieved from the [Inflammatory Bowel Disease Multi'omics Database] (<https://ibdmdb.org/tunnel/public/HMP2/16S/1806/products>).  

```{r microbiome count data}

# read in 16S abundance/ count data in HMP2 database 
untar("hmp2_abund.tar", exdir = getwd()) # unzip a .tar file 

abund_raw <- read_tsv("taxonomic_profiles.tsv.gz", 
                      show_col_types = FALSE) %>% 
        as.data.frame() 

``` 

Read in the metadata in `.csv` format. 
```{r metadata dataframe}
# read in the meta data 
meta_raw <- read.csv("hmp2_metadata.csv")

```

Subset identifiers, environmental/ clinical variables of my research interest, including `Project`, `External.ID`, `Participant.ID`, `week_num`, `biopsy_location`, `diagnosis` and `Age.at.diagnosis`. 

```{r subset env variables}

meta_raw_sub <- meta_raw %>% 
        select(Project, 
               External.ID,
               Participant.ID,
               diagnosis,
               week_num, 
               biopsy_location,
               Age.at.diagnosis) 

```

Merge 16S abundance data and subset metadata. 
```{r merge 16S and metadata}

# reshape abundance data 
abund_raw_id <- abund_raw %>% 
        gather(key="sample_id",
               value = "read",
               -c("#OTU ID", "taxonomy")) 

# join two data frames by "id" = "External.ID" 
abund_meta <- abund_raw_id %>% 
        inner_join(meta_raw_sub, 
                   by = c("sample_id" = "External.ID"))

```


The microbiome 16S abundance data set contains data points from `r length(unique(abund_meta$Participant.ID))` participants; the biopsy were collected from 7 different locations, including `Rectum`, `Ileum`, `Sigmoid Colon`, `Descending (left-sided) colon`, `Cecum`, `Ascending (right-sided) colon`, and `Transverse colon`. \ 
For the simplicity purpose, I focused on rectum biopsy data for the subsequent analysis. 
```{r extract only rectum biopsy data points} 

abund_meta_rectum <- abund_meta %>% 
        filter(biopsy_location == "Rectum") 

```


Collapse taxonomy at the genus level, namely similarity equals or higher than 97% 
```{r collapse the taxonomy at the genus level}

abund_meta_rectum2 <- abund_meta_rectum %>%
        mutate(genus = str_split_fixed(abund_meta_rectum$taxonomy, "; __", 6)[,6]) %>%  
        mutate(genus_rep = case_when(genus == "g" ~ "unclassified", 
                                     genus != "g" ~ genus)) %>%
        select(-genus)

```


Select columns, `sample_id`, `Project`, `Participant.ID`, `read`, `diagnosis` and `genus_rep` from the above data frame. And then calculate total reads for each individual as the offset. 

```{r offset calculation}

abund_meta_rectum3 <- abund_meta_rectum2 %>% 
        select(sample_id, Participant.ID, read, diagnosis, genus_rep) %>%
        rename(subject_id = Participant.ID)

df_total_reads <- abund_meta_rectum3 %>% 
        group_by(sample_id) %>% 
        summarise(offset = sum(read)) %>%
        ungroup() 

abund_meta_rectum4 <- abund_meta_rectum3 %>% 
        left_join(df_total_reads, by="sample_id") 


# convert to the factor variable with new levels 
abund_meta_rectum4$diagnosis <- factor(abund_meta_rectum4$diagnosis,
                                       levels = c("nonIBD", "CD", "UC"))


# abund_meta_rectum3 %>% 
#         group_by(sample_id) %>%
#         summarise(lib_size= sum(read)) %>% 
#         arrange(lib_size) %>%
#         ggplot(aes(x=1, y=lib_size)) +  # visualize distribution of variations in library sizes 
#         geom_jitter()+
#         scale_y_log10() 

```
The sampling depth varied by samples which likely affect hypothesis analysis. \ 
I consider to choose a best depth threshold and to conduct rarefaction, following the [instruction] (<https://www.youtube.com/watch?v=c7H8jLjTxSE>) introduced by Dr. Pat Schloss. \ 
The optimal sampling depth threshold is determined according to the good's coverage statistics. 

```{r determin best sampling depth}

# abund_meta_rectum3 %>% 
#         group_by(sample_id) %>% 
#         summarise(singles = sum(read == 1),
#                   lib_size = sum(read)) %>% 
#         mutate(goods_cov = 100*(1-singles/lib_size)) %>%  # calculate the good's coverage 
#         ggplot(aes(x= lib_size, y=goods_cov))+ # plot sampling size by goods coverage 
#         geom_point() + 
#         geom_hline(yintercept = 97, color = "red")  # good's coverage = 97 percentage 
# 
# 
# abund_meta_rectum3_goods <- abund_meta_rectum3 %>% 
#         group_by(sample_id) %>% 
#         summarise(singles = sum(read == 1),
#                   lib_size = sum(read)) %>% 
#         mutate(goods_cov = 100*(1-singles/lib_size)) 
# 
# 
# abund_meta_rectum3_goods %>% 
#         filter(goods_cov >= 97) %>% 
#         arrange(lib_size)  # sampling depth threshold = 286 reads
# 
# 
# # identify ids for samples below the minimal sampling depth 
# sample_drop <- abund_meta_rectum3_goods %>% 
#         filter(goods_cov <97)  %>% 
#         pull(sample_id)
# 

```


## Step 2 Investigate outcome distribution zeros 

Subset samples with zero percentage higher ranging from 50% to close 100%. Apply models specializing in the zero-inflated data to the sub-samples. 
```{r histogram of zero distribution in genus}

df_zero_percent <- abund_meta_rectum4 %>% 
        filter(genus_rep != "unclassified") %>% 
        group_by(genus_rep) %>% 
        summarise(zero_percent = mean(read == 0)) 

df_zero_percent$zero_cut <- cut(df_zero_percent$zero_percent, breaks = 4)  # cut zero_percentage into four groups 

levels(df_zero_percent$zero_cut)

# subset variables for the genus with many zero 
abund_meta_rectum5 <- abund_meta_rectum4 %>% 
        left_join(df_zero_percent, by= "genus_rep") %>% 
        arrange(desc(zero_percent)) 

```


Create histogram for representative genus, `_Clostridium_innocuum_group` with excessive zeros 
```{r histogram for zero distribution}

abund_meta_rectum5 %>% 
        filter(zero_percent != 1) %>% 
        filter(zero_cut == "(0.506,0.753]") %>% 
        head(10)


# use "_Clostridium_innocuum_group" as an example for plotting histogram 
abund_meta_rectum5 %>% 
        filter(genus_rep == "_Clostridium_innocuum_group") %>% 
        ggplot(aes(x=read))+
         geom_histogram(binwidth = 15)+
        xlab("Observed read values")+
        ylab("Frequency") 


```


## Step 3 Create offset for poisson/ NB models to calculate ratios


## Step 4 Create a formula for fitting ZIP/ ZINB and ZHP/ ZHNB

## Step 5 Fit ZIP/ ZINB and ZHP/ ZHNB 

## Step 6 Select the optimal model 

## Step 7 Interpret models' output 











